File: 1138-grumpy-bookstore-owner/grumpy-bookstore-owner.cpp
class Solution {
public:
    int maxSatisfied(vector<int>& customers, vector<int>& grumpy, int minutes) 
    {
        int n = customers.size(), satisfaction = 0, maxExtra = 0, extra = 0;
        for(int i = 0; i < n; i++)
        {
            if(!grumpy[i])
            satisfaction += customers[i];

            else
            extra += customers[i];

            if(i - minutes + 1 >= 0)
            {
                maxExtra = max(maxExtra, extra);
                if(grumpy[i - minutes + 1])
                extra -= customers[i - minutes + 1];
            }
        }
        return satisfaction + maxExtra;
    }
};

File: 1138-grumpy-bookstore-owner/grumpy-bookstore-owner.java
class Solution {
    public int maxSatisfied(int[] cust, int[] grumpy, int minutes) 
    {
        int satisfiedCustomers=0;
        int maxExtraSatisfiedCustomers=0;
        int extraSatisfiedCustomers=0;
        int i=0;
        for(int j=0;j<cust.length;j++)
        {
            if(grumpy[j]==0)
            satisfiedCustomers+=cust[j];

            else                                          //(grumpy[j]==1)
            extraSatisfiedCustomers+=cust[j];

            if(j-i+1==minutes)
            {
                maxExtraSatisfiedCustomers=Math.max(maxExtraSatisfiedCustomers,extraSatisfiedCustomers);
                if(grumpy[i]==1)
                extraSatisfiedCustomers-=cust[i];
                i++;
            }
        }

        satisfiedCustomers+=maxExtraSatisfiedCustomers;
        return satisfiedCustomers;
    }
}